library/psa_crypto.c

```diff
@@ -523,283 +504,348 @@ static psa_status_t psa_check_rsa_key_byte_aligned(
     mbedtls_mpi_free( &n );
     return( status );
 }
+#endif /* MBEDTLS_PK_PARSE_C */
 
-static psa_status_t psa_import_rsa_key( psa_key_type_t type,
-                                        const uint8_t *data,
-                                        size_t data_length,
-                                        mbedtls_rsa_context **p_rsa )
+/* Convert import representation to pk_context containing an RSA key.
+ * Note: upon return of PSA_SUCCESS, the caller is responsible for freeing
+ * the PK context object. */
+static psa_status_t psa_load_rsa_representation( const psa_key_slot_t *slot,
+                                                 mbedtls_pk_context *ctx )
 {
+#if defined(MBEDTLS_PK_PARSE_C)
     psa_status_t status;
-    mbedtls_pk_context pk;
-    mbedtls_rsa_context *rsa;
     size_t bits;
-
-    mbedtls_pk_init( &pk );
+    mbedtls_pk_init( ctx );
 
     /* Parse the data. */
-    if( PSA_KEY_TYPE_IS_KEY_PAIR( type ) )
+    if( PSA_KEY_TYPE_IS_KEY_PAIR( slot->attr.type ) )
         status = mbedtls_to_psa_error(
-            mbedtls_pk_parse_key( &pk, data, data_length, NULL, 0 ) );
+            mbedtls_pk_parse_key( ctx, slot->data.key.data, slot->data.key.bytes, NULL, 0 ) );
     else
         status = mbedtls_to_psa_error(
-            mbedtls_pk_parse_public_key( &pk, data, data_length ) );
+            mbedtls_pk_parse_public_key( ctx, slot->data.key.data, slot->data.key.bytes ) );
+
     if( status != PSA_SUCCESS )
         goto exit;
 
     /* We have something that the pkparse module recognizes. If it is a
      * valid RSA key, store it. */
-    if( mbedtls_pk_get_type( &pk ) != MBEDTLS_PK_RSA )
+    if( mbedtls_pk_get_type( ctx ) != MBEDTLS_PK_RSA )
     {
         status = PSA_ERROR_INVALID_ARGUMENT;
         goto exit;
     }
 
-    rsa = mbedtls_pk_rsa( pk );
     /* The size of an RSA key doesn't have to be a multiple of 8. Mbed TLS
      * supports non-byte-aligned key sizes, but not well. For example,
      * mbedtls_rsa_get_len() returns the key size in bytes, not in bits. */
-    bits = PSA_BYTES_TO_BITS( mbedtls_rsa_get_len( rsa ) );
+    bits = PSA_BYTES_TO_BITS( mbedtls_rsa_get_len( mbedtls_pk_rsa( *ctx ) ) );
     if( bits > PSA_VENDOR_RSA_MAX_KEY_BITS )
     {
         status = PSA_ERROR_NOT_SUPPORTED;
         goto exit;
     }
-    status = psa_check_rsa_key_byte_aligned( rsa );
+    status = psa_check_rsa_key_byte_aligned( mbedtls_pk_rsa( *ctx ) );
 
 exit:
-    /* Free the content of the pk object only on error. */
     if( status != PSA_SUCCESS )
+        mbedtls_pk_free( ctx );
+
+    return( status );
+#else
+    (void) slot;
+    (void) ctx;
+    return( PSA_ERROR_NOT_SUPPORTED );
+#endif /* MBEDTLS_PK_PARSE_C */
+}
+
+static psa_status_t psa_export_rsa_key( psa_key_type_t type,
```

Comment:
This function needs some documentation. What does it do if `pk` is not an RSA key? Why does it take an `mbedtls_pk_context` argument and not an `mbedtls_rsa_context` anyway? And likewise for `psa_load_rsa_representation`. The RSA parsing and writing functions need a pk context, but that's an implementation detail, so I think it would make more sense to take an `mbedtls_rsa_context*` argument and define an `mbedtls_pk_context` variable locally.

Reply:
Updated